File: code/threading.Timer2/python/timer_cancel.py
#!/usr/bin/env python3

import threading
import datetime
import time

def print_message(message):
    print(f"{datetime.datetime.now()}: {message}")

def say_hello():
    print_message("Hello, World!")

def main():
    timer = threading.Timer(5.0, say_hello)
    timer.start()
    print_message("Timer started, waiting for it to trigger...")
    
    time.sleep(1)
    print_message("Now we will cancel it.")
    timer.cancel()
    
    while timer.is_alive():
        print_message("Timer is active...")
        time.sleep(1)
    
    print_message("Timer cancelled.")

if __name__ == "__main__":
    main()

File: code/timer2/recipe-01/timer_cancel.cpp
#include "timer.hpp"
#include "print_message.hpp"
#include <chrono>
#include <thread>

void say_hello() {
    print_message("Hello, World!");
}

int main() {
    Timer timer;
    timer.start(say_hello, std::chrono::seconds{3});
    print_message("Timer started, waiting for it to trigger...");

    std::this_thread::sleep_for(std::chrono::seconds{1});
    print_message("Now we will cancel it.");
    timer.cancel();

    while (timer.isActive()) {
        print_message("Timer is active...");
        std::this_thread::sleep_for(std::chrono::seconds{1});
    }
    print_message("Timer cancelled.");
}

